src/rules/allPropertiesHaveExplicitDefined/allPropertiesHaveExplicitDefined.ts

```diff
@@ -0,0 +1,262 @@
+                            continue;
+                        }
+                        const type = getType(
+                            propertyDefinition.typeAnnotation.typeAnnotation,
+                            esTreeNodeToTSNodeMap,
+                            checker
+                        );
+                        if (
+                            propertyDefinition.optional ||
+                            isNullableType(type)
+                        ) {
+                            if (!decoratorsStatus.isOptional) {
+                                context.report({
+                                    node: propertyDefinition,
+                                    messageId: "missing-is-optional-decorator",
+                                });
+                            }
+                        } else {
+                            if (
+                                !decoratorsStatus.isDefined &&
+                                !decoratorsStatus.hasTypeCheckingDecorator
+                            ) {
+                                context.report({
+                                    node: propertyDefinition,
+                                    messageId: "missing-is-defined-decorator",
+                                });
+                            }
+                        }
+                    }
+                }
+            },
+        };
+    },
+});
+
+export default rule;
+
+type DecoratorsStatus = {
+    isDefined: boolean;
+    hasTypeCheckingDecorator: boolean;
+    isOptional: boolean;
+};
+
+function getType(
+    typeNode: TypeNode,
+    esTreeNodeToTSNodeMap: ParserWeakMapESTreeToTSNode,
+    checker: TypeChecker
+): Type {
+    const tsNode = esTreeNodeToTSNodeMap.get(typeNode);
+    const type = checker.getTypeAtLocation(tsNode);
+    return type;
+}
+
+function getDecoratorsStatus(
+    propertyDefinition: PropertyDefinition
+): DecoratorsStatus {
+    let isDefined = false;
+    let hasTypeCheckingDecorator = false;
+    let isOptional = false;
+    if (propertyDefinition.decorators) {
+        for (const decorator of propertyDefinition.decorators) {
+            if (
+                decorator.expression.type === AST_NODE_TYPES.CallExpression &&
+                decorator.expression.callee.type === AST_NODE_TYPES.Identifier
+            ) {
+                if (decorator.expression.callee.name === "IsDefined") {
+                    isDefined = true;
+                }
+                if (
+                    TYPE_CHECKING_DECORATOR.has(
+                        decorator.expression.callee.name
+                    )
+                ) {
+                    hasTypeCheckingDecorator = true;
+                }
+                if (decorator.expression.callee.name === "IsOptional") {
+                    isOptional = true;
+                }
+            }
+        }
+    }
+    return {isDefined, hasTypeCheckingDecorator, isOptional};
+}
+
+const TYPE_CHECKING_DECORATOR = new Set([
```

Comment:
this set appears three times in the code base - there is an import from the class-validator library, a list i created that is an array, and now this set. 

if they all contain the same items it might be best to have it in just one place?